### ISSUE TITLE
Simplify Testing System for Rules

### ISSUE THREAD

**Issue by JKillian** [palantir/tslint#620]
TODO (based off of new testing system explained [here](https://github.com/palantir/tslint/issues/620#issuecomment-144561726)):

- [x] Write code to run new-style tests
- [x] Write tests for code that runs new-style tests
- [x] Convert old-style tests into new-style tests for each rule
- [x] Clean up all of the above
- [ ] Document how to write tests on website
- [x] Make `--test` command so custom rules can be tested the same way?

---
Original Post:
Finding numerical positions of failures is tedious and inflexible (if the test file changes, all the failures move.) Instead, we could use some sort of marking system:

```typescript
/*fail:0*/var/*endfail:0*/ foo;

function tmp(t: any) {
    /*fail:1*/var/*endfail:1*/ x = 3;
}

/*fail:2*/var/*endfail:2*/ i,
    j;

/*fail:3*/var/*endfail:3*/ [a, b] = [1, 2];

declare var tmp2: any;

let bar;
const qux;

let k,
    l;

module quz {
    export var i;
}

let [x, y] = [1, 2];
```

Admittedly this makes the test files slightly harder to read, but overall it seems like a much easier and maintainable system. (This issue is about coming up with a better system in general, it doesn't have to follow the example above)

**Comment by ashwinr** [comment]
I strongly disagree -- this is an unnecessary complication and makes the test files unreadable, sprinkled with magical markers. I would rather allow reading code off a string buffer and not mandate making a separate test file for each test.

**Comment by JKillian** [comment]
Hmm, each test essentially already has a separate source file though, in `test/files/rules`, as well as well as the file that runs the test in `test/rules`.  The files in `test/rules` are essentially unreadable now:

```typescript
describe("<no-trailing-whitespace>", () => {
    const NoTrailingWhitespaceRule = Lint.Test.getRule("no-trailing-whitespace");
    const fileName = "rules/trailing.test.ts";
    const createFailure = Lint.Test.createFailuresOnFile(fileName, NoTrailingWhitespaceRule.FAILURE_STRING);

    it("forbids trailing whitespace", () => {
        const actualFailures = Lint.Test.applyRuleOnFile(fileName, NoTrailingWhitespaceRule);
        // Ahh! Lots of random numbers!!!!
        const expectedFailure1 = createFailure([2, 24], [2, 28]);
        const expectedFailure2 = createFailure([3, 32], [3, 36]);
        const expectedFailure3 = createFailure([5, 1], [5, 5]);
        const expectedFailure4 = createFailure([6, 1], [6, 5]);
        const expectedFailure5 = createFailure([9, 2], [9, 6]);

        Lint.Test.assertContainsFailure(actualFailures, expectedFailure1);
        Lint.Test.assertContainsFailure(actualFailures, expectedFailure2);
        Lint.Test.assertContainsFailure(actualFailures, expectedFailure3);
        Lint.Test.assertContainsFailure(actualFailures, expectedFailure4);
        Lint.Test.assertContainsFailure(actualFailures, expectedFailure5);
        assert.lengthOf(actualFailures, 5);
    });
});
```

This also wouldn't disallow reading from a string buffer - input could be read from a file or from a string, whichever was preferred.

**Comment by ashwinr** [comment]
discussed offline -- let's support multi-line string templates for test code instead.

**Comment by RyanCavanaugh** [comment]
Any updates on this? I want to write a rule but the current test system is like pulling teeth (`expected 7 to equal 14`? Oh, well in that case...).

My 2 cents on the design is that TypeScript used to have separate files full of hardcoded positional values, as exists today in TSLint, and it was an unmitigated maintainability disaster. We moved some tests over to be multi-line strings inside the test files accompanied by more hardcoded positional values, and that was a barely-mitigated maintainability disaster.

What you need is a moderately expressive DSL that lets you mix "sample code" and its metadata seamlessly in a *readable* format, without tons of boilerplate in each unit test file (I had to copy and paste 12 lines of unit test code to replace 4 small strings when adding my rule test -- not a good sign!).

I would write something like this:
**curly-rule-test-1.ts** (this name maps to `curly-rule`; multiple files may exist)
```ts
if (foo)
  bar;
  ~~~~ IF_FAILURE_STRING 

for(foo; bar; baz) expr;
                   ~~~~~ FOR_FAILURE_STRING
```
In this system, a line consisting of only whitespace and ~, optionally followed by an identifier, represents an expected error indicator line. It's rather easy to take a file like this and produce spans that you can compare against the actual output of the linter, or conversely to take a set of linting errors and produce a file that shows where the error spans are (which is incredibly helpful when debugging).

When we moved our baseline reporter in TypeScript to do something like this, our accuracy in catching error quality regressions went *way* up.

Now instead of an error like `expected 14 to equal 11` you can issue an error like:
```ts
in answers-only-rule.ts, expected
    42 + 12;
    ~~ ONLY_FORTYTWO_ALLOWED_STRING
to equal
    42 + 12;
         ~~ ONLY_FORTYTWO_ALLOWED_STRING
```

**Comment by JKillian** [comment]
:+1:  The current system is pretty (very?) broken, it takes a good amount of boilerplate to run any tests, then figuring out the positions is awful, and figuring out what the Rule is actually doing correctly/incorrectly also is painful.

@RyanCavanaugh If I have the time, I want to put together and prototype something like what you suggested (I had forgotten about this at the moment honestly). In your proposal, what would you suggest for errors that span multiple lines?

**Comment by RyanCavanaugh** [comment]
I would write it like this, by analogy of the C preprocessor's line continuation character:

```ts
function bleh() {
~~~~~~~~~~~~~~~~~\
  throw null;
~~~~~~~~~~~~~\
}
~ FUNCTION_MUST_BE_COOL
```
Alternatively, just say that any line consisting solely of `~` is necessarily continued on the next line -- if all errors must be named by a `~ NAME` sequence, then that continuation is implicit. A test parse error occurs if there is a non-contiguous sequence of `~` isn't followed by an identifier.

**Comment by JKillian** [comment]
## Rules Testing System Proposal ##

#### Goals ####
Make writing tests for rules all around better. This includes:

* Reducing brittleness of tests
* Reducing boilerplate needed to add tests
* Improving test code readability/clarity (it's hard to tell what's actually being tested now)
* Improving test result clarity (it's hard to tell actual vs. expected)
* Expose testing infrastructure for custom rules

#### Usage ####
The current TSLint CLI syntax is succinct for linting files and while this new system will be part of a major TSLint release (and thus can contain breaking changes) I don't feel that tacking on additional functionality to the CLI would be that great.
Instead, a separate binary, `tslint-test` can be included with TSLint specifically for running tests. 

```
tslint-test tests-directory [rule-files...]
```

TSLint would use this as part of its build to run tests on its rules. Those developing custom rules could run their tests manually with the tool or integrate it into a build.

#### Structure ####

The rules provided to `tslint-test` would be a list of compiled `.js` rules following the same structure as current TSLint rules.

The tests directory would contain a subdirectory for each rule being tested. So if `MaxLineLengthRule.js` is being tested, the tests directory should contain a directory called  `MaxLineLengthTests`. *Should these directories be CapitalCase or hyphen-case or snake_case?*

Each individual directory can have test cases with any name ending with a `.test.ts`,`.test.tsx`, or `.test.d.ts` extension. The corresponding rule will be run on each of these test files independently of each other. Occasionally, a test might need to consist of multiple files. In this case, a directory ending with `TestGroup` can be created. Test files in these directories will be processed together by TSLint.

Here's a complete example:

```
PWD:
├── myTestsDirectory
│   ├── FindCrazyBugsTests
│   │   ├── data-race.test.ts
│   │   ├── memory-leak.test.tsx
│   │   └── shellshock.test.d.ts
│   ├── ImportTests
│   │   ├── ImportOthersTestGroup
│   │   │   ├── importer.test.ts
│   │   │   └── importme.test.ts
│   │   └── import-builtins.test.ts
│   └── SomethingCoolTests
│       ├── coolness.test.ts
│       └── ensure-chill.test.tsx
├── findCrazyBugsRule.js
├── importRule.js
└── somethingCoolRule.js

tslint-test Usage:
tslint-test myTestsDirectory findCrazyBugsRule.js somethingCoolRule.js
```

#### Test Syntax ####

Tests files have a simple DSL consisting of `ts` or `tsx` code with additional markup that indicates errors. The markup indicates the span for which a rule should detect an error and the error message the rule should produce.

Say we have a rule that produces an error anytime a function named `evil` is declared. The rule's error message is "Functions named evil are evil". Here's an example of a test for this rule:


```
function evil() {
         ~~~~                             [Functions named evil are evil]
  console.log("This function is evil!");
}
```

The syntax for marking an error on a line is, informally, a line of tildes underneath the span where the error occurred followed by optional whitespace followed by the error message enclosed in brackets. A valid line matches this reggae:

```
/^\s*~+\s*\[[\w ]+\]\s*$/

```

###### Multiple Errors on the Same Line ######
Occasionally a rule may need to detect multiple errors on the same line. In this case, errors can be specified on subsequent lines following the line of code with the errors.

Say we have a rule that disallows the use of `2` as a numeric literal:

```
let x = 2 + 3 + 2;
        ~          [Use of numeric literal 2 illegal]
                ~  [Use of numeric literal 2 illegal]
```

###### Errors Spanning Multiple Lines ######
Errors might need to span multiple lines. To indicate this, the error string is simply not specified until the last line of the error. 

For example, say we have a rule which disallows the use of `while` statements:


    while(true) {
    ~~~~~~~~~~~~~
      computeLotteryWinners();
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
    }
    ~ [Use of while illegal]                        


###### Nested Multiline Errors ######
If errors are nested and span multiple lines, the order of markup lines indicates what error the line goes with. 

Using the same rule about `while` statements:

    while(true) {
    ~~~~~~~~~~~~~
      while(5 > 3) {
    ~~~~~~~~~~~~~~~~
      ~~~~~~~~~~~~~~
        computeLotteryWinners();
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      }
    ~~~
    ~~~ [Use of while illegal]      
    }
    ~ [Use of while illegal]                        


###### Possible Feature: Error Message Shorthand ######
It might be annoying to copy and paste the same error messages everywhere. A few alternatives:

**Possibility 1: Per-file 'variables' for error messages:**

```
console.log("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~ [0]
console.log("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~ [0]
console.warn("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~~ [1]

{0}: console.log is disallowed
{1}: console.warn is disallowed
```

**Possibility 2: Per-directory 'variables for error messages:**

*TestDir/NoConsoleTests/log.test.ts*

```
console.log("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~ [0]
console.log("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~ [0]
console.warn("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~~ [1]
```

*TestDir/NoConsoleTests/errors.txt*

```                
{0}: console.log is disallowed
{1}: console.warn is disallowed
```

**Possibility 3: Rules each have an CreateErrorMessage method**

Each rule would have a create error message method that created the error message for the rule based off some amount of arguments. Test files would just specify these arguments for each error.

```
console.log("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~ ["console.log"]
console.log("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~ ["console.log"]
console.warn("whatever");
~~~~~~~~~~~~~~~~~~~~~~~~~ ["console.warn"]
```

#### Processing Tests ####
`tslint-test` reads the markup lines from test files and stores spans and messages for errors in memory. It creates a temporary version of the files with markup removed and passes these to TSLint. 

#### Output ####
[TBD]

**Comment by JKillian** [comment]
Thanks @RyanCavanaugh, I wrote out a more detailed proposal based off of your ideas. Any feedback on parts that seem good/bad? @adidahiya, @ashwinr also

**Comment by RyanCavanaugh** [comment]
:+1: :+1:

I like *Possibility 1: Per-file 'variables' for error messages*. Putting stuff in other files is bad for reviewing changes in e.g. GitHub so I don't like possibility 2.

**Comment by JKillian** [comment]
Finally getting around to implementing this. A very messy (but working) version is on the `new-test-system` branch.  Yay!

![](http://imgur.com/bxztpGY.png)

(In this screenshot, I made the `.linttest` file intentionally wrong so we could see an error - normally it would be the other way around)

**Comment by JKillian** [comment]
By the way, if anyone wants to help convert tests it would be much appreciated! It's quite easy, [see this example commit](https://github.com/palantir/tslint/commit/5756ca7860d3cb0a83754ed4c013714dd6e5d608). It's just takes awhile to do with one person for all the rules. 

Running `grunt test` will run the new tests you add along with the other normal tslint tests.

**Comment by JKillian** [comment]
Finished convert the tests for each rule over to the new system, next up is cleaning up code and documentation.

**Comment by JKillian** [comment]
Remaining documentation task noted in #941
